lib/elixir/src/elixir_errors.erl

```diff
@@ -30,25 +30,32 @@ form_error(Meta, File, Module, Desc) ->
   compile_error(Meta, File, Module:format_error(Desc)).
 
 -spec form_warn(list(), binary() | #{file := binary()}, module(), any()) -> ok.
-form_warn(Meta, File, Module, Desc) when is_list(Meta), is_binary(File) ->
-  do_form_warn(Meta, File, #{}, Module:format_error(Desc));
-form_warn(Meta, #{file := File} = E, Module, Desc) when is_list(Meta) ->
-  do_form_warn(Meta, File, E, Module:format_error(Desc)).
+form_warn(Meta, E, Module, Desc) ->
+  {Line, File, Warning, Message} = format_form_warn(Meta, E, Module, Desc),
+  io_warn(Line, File, Warning, Message).
 
-do_form_warn(Meta, GivenFile, E, Warning) ->
-  {File, Line} = meta_location(Meta, GivenFile),
+-spec format_form_warn(list(), binary() | #{file := binary()}, module(), any()) -> {integer(), binary(), any(), unicode:chardata()}.
+format_form_warn(Meta, E, FormatModule, Desc) when is_list(Meta) ->
```

Comment:
I would prefer to pass File, Module, Function as explicit arguments instead of `E` being two different shapes. WDYT?

Reply:
I followed the existing `form_warn` functionality. I can refactor this so the env arguments are passed explicitly but it would touch everywhere we use `form_warn` so I think it should be a separate PR.